File: functions/sandbox-node/handler.ts
/**
 * Node sandbox Lambda for workspace file execution.
 * Keep Node runtime process execution here.
 */

import { access, readdir, stat } from "node:fs/promises";
import { createHash } from "node:crypto";
import { readFile, writeFile } from "node:fs/promises";
import { basename, dirname, relative, resolve } from "node:path";
import { spawn } from "node:child_process";
import process from "node:process";
import ts from "typescript";
import type { Stats } from "node:fs";

interface SandboxRequest {
  runtime: "node";
  namespace: string;
  entryPath: string;
  args?: string[];
  workspaceRoot?: string;
  timeoutSeconds: number;
  outputLimitBytes: number;
}

interface SandboxResponse {
  ok: boolean;
  runtime: "node";
  exitCode: number | null;
  stdout: string;
  stderr: string;
  artifacts?: SandboxArtifact[];
  durationMs: number;
  timedOut?: boolean;
  truncated?: boolean;
}

interface SandboxArtifact {
  kind: "file";
  path: string;
  mediaType: string;
  dataBase64: string;
  metadata: {
    size: number;
  };
}

interface WorkspaceSnapshot {
  files: Map<string, { mtimeMs: number; size: number }>;
}

const MAX_ARTIFACT_BYTES = 256 * 1024;
const ENTRY_FILE_WAIT_MS = 5000;
const ENTRY_FILE_WAIT_INTERVAL_MS = 100;

export async function handler(event: SandboxRequest): Promise<SandboxResponse> {
  const startedAt = Date.now();

  try {
    if (event.runtime !== "node") {
      throw new Error("sandbox-node only supports node requests");
    }
    if (!event.entryPath.endsWith(".js") && !event.entryPath.endsWith(".ts")) {
      throw new Error("node sandbox only executes .js and .ts files");
    }

    const workspaceRoot = resolveWorkspaceRoot(event.workspaceRoot, event.namespace);
    const filePath = resolveWorkspacePath(workspaceRoot, event.entryPath);
    await waitForFile(filePath);
    const before = await snapshotWorkspace(workspaceRoot);
    const executablePath = event.entryPath.endsWith(".ts") ? await prepareTypescriptEntry(filePath, workspaceRoot) : filePath;
    const result = await runNodeFile(executablePath, workspaceRoot, event.args ?? [], event.timeoutSeconds, event.outputLimitBytes);
    return {
      ...result,
      runtime: "node",
      artifacts: await collectChangedArtifacts(workspaceRoot, before),
      durationMs: Date.now() - startedAt,
    };
  } catch (err) {
    return {
      ok: false,
      runtime: "node",
      exitCode: null,
      stdout: "",
      stderr: err instanceof Error ? err.message : String(err),
      durationMs: Date.now() - startedAt,
    };
  }
}

async function prepareTypescriptEntry(filePath: string, workspaceRoot: string): Promise<string> {
  const source = await readFile(filePath, "utf8");
  const transpiled = ts.transpileModule(source, {
    compilerOptions: {
      module: ts.ModuleKind.ESNext,
      target: ts.ScriptTarget.ES2022,
      esModuleInterop: true,
      sourceMap: false,
    },
    fileName: filePath,
  }).outputText;
  const generatedPath = resolveTypescriptOutputPath(filePath, workspaceRoot, source);
  await writeFile(generatedPath, transpiled, "utf8");
  return generatedPath;
}

async function waitForFile(filePath: string): Promise<void> {
  const deadline = Date.now() + ENTRY_FILE_WAIT_MS;
  let lastError: unknown;

  while (Date.now() < deadline) {
    try {
      await access(filePath);
      return;
    } catch (err) {
      lastError = err;
      await new Promise((resolve) => setTimeout(resolve, ENTRY_FILE_WAIT_INTERVAL_MS));
    }
  }

  if (lastError) {
    throw lastError;
  }
  await access(filePath);
}

async function snapshotWorkspace(root: string): Promise<WorkspaceSnapshot> {
  const files = new Map<string, { mtimeMs: number; size: number }>();
  await walkWorkspace(root, async (path, stats) => {
    files.set(relative(root, path), {
      mtimeMs: stats.mtimeMs,
      size: stats.size,
    });
  });
  return { files };
}

async function collectChangedArtifacts(root: string, before: WorkspaceSnapshot): Promise<SandboxArtifact[]> {
  const artifacts: SandboxArtifact[] = [];
  await walkWorkspace(root, async (path, stats) => {
    const relativePath = relative(root, path);
    const previous = before.files.get(relativePath);
    if (previous?.mtimeMs === stats.mtimeMs && previous.size === stats.size) {
      return;
    }
    if (stats.size > MAX_ARTIFACT_BYTES) {
      return;
    }

    const content = await readFile(path);
    artifacts.push({
      kind: "file",
      path: `/${relativePath}`,
      mediaType: "application/octet-stream",
      dataBase64: content.toString("base64"),
      metadata: {
        size: stats.size,
      },
    });
  });
  return artifacts;
}

async function walkWorkspace(
  root: string,
  visit: (path: string, stats: Stats) => Promise<void>,
): Promise<void> {
  for (const entry of await readdir(root, { withFileTypes: true })) {
    if (entry.name === "." || entry.name === "..") {
      continue;
    }

    const path = resolve(root, entry.name);
    const stats = await stat(path);
    if (stats.isDirectory()) {
      await walkWorkspace(path, visit);
      continue;
    }
    if (stats.isFile()) {
      await visit(path, stats);
    }
  }
}

function resolveTypescriptOutputPath(filePath: string, workspaceRoot: string, source: string): string {
  const fileName = basename(filePath).replace(/\.ts$/, "");
  const hash = createHash("sha256").update(filePath).update("\0").update(source).digest("hex").slice(0, 16);
  const generated = resolve(dirname(filePath), `.${fileName}-${hash}.js`);
  const relation = relative(workspaceRoot, generated);
  if (relation.startsWith("..") || relation === ".." || relation === "" || relation.startsWith("/")) {
    throw new Error("Invalid generated entry path: resolved outside workspace root");
  }
  return generated;
}

async function runNodeFile(
  filePath: string,
  cwd: string,
  args: string[],
  timeoutSeconds: number,
  outputLimitBytes: number,
): Promise<Omit<SandboxResponse, "runtime" | "durationMs">> {
  const child = spawn(process.execPath, [filePath, ...args], {
    cwd: cwd,
    env: {
      PATH: "/usr/local/bin:/usr/bin:/bin",
      HOME: "/tmp",
      TMPDIR: "/tmp",
      NODE_OPTIONS: "--no-deprecation",
    },
    stdio: ["ignore", "pipe", "pipe"],
  });

  return collectChildResult(child, timeoutSeconds, outputLimitBytes);
}

function resolveWorkspaceRoot(root: string | undefined, namespace: string): string {
  assertSafeNamespace(namespace);
  return resolve(root || "/mnt/workspaces", namespace);
}

function resolveWorkspacePath(workspaceRoot: string, entryPath: string): string {
  const normalizedEntry = entryPath.startsWith("/") ? entryPath.slice(1) : entryPath;
  const resolved = resolve(workspaceRoot, normalizedEntry);
  const relation = relative(workspaceRoot, resolved);
  if (relation.startsWith("..") || relation === ".." || relation === "" || relation.startsWith("/")) {
    throw new Error("Invalid entry path: resolved outside workspace root");
  }
  return resolved;
}

function assertSafeNamespace(namespace: string): void {
  if (!/^fs-[a-f0-9]{40}$/.test(namespace)) {
    throw new Error("Invalid workspace namespace");
  }
}

async function collectChildResult(
  child: ReturnType<typeof spawn>,
  timeoutSeconds: number,
  outputLimitBytes: number,
): Promise<Omit<SandboxResponse, "runtime" | "durationMs">> {
  const stdoutChunks: Buffer[] = [];
  const stderrChunks: Buffer[] = [];
  child.stdout?.on("data", (chunk: Buffer) => stdoutChunks.push(chunk));
  child.stderr?.on("data", (chunk: Buffer) => stderrChunks.push(chunk));

  let timedOut = false;
  const timeout = setTimeout(() => {
    timedOut = true;
    child.kill("SIGKILL");
  }, timeoutSeconds * 1000);

  const exitCode = await new Promise<number | null>((resolve) => {
    child.on("close", (code) => resolve(code));
  });
  clearTimeout(timeout);

  const stdout = truncateOutput(Buffer.concat(stdoutChunks), outputLimitBytes);
  const stderr = truncateOutput(Buffer.concat(stderrChunks), outputLimitBytes);
  return {
    ok: exitCode === 0 && !timedOut,
    exitCode,
    stdout: stdout.value,
    stderr: timedOut ? appendTimeout(stderr.value, timeoutSeconds) : stderr.value,
    timedOut,
    truncated: stdout.truncated || stderr.truncated,
  };
}

function truncateOutput(value: Buffer, limit: number): { value: string; truncated: boolean } {
  if (value.byteLength <= limit) {
    return { value: value.toString("utf8"), truncated: false };
  }

  return {
    value: `${value.subarray(0, limit).toString("utf8")}\n[output truncated]`,
    truncated: true,
  };
}

function appendTimeout(stderr: string, timeoutSeconds: number): string {
  return [stderr, `Timed out after ${timeoutSeconds}s`].filter(Boolean).join("\n");
}

File: functions/sandbox-python/handler.py
"""
Python sandbox Lambda for workspace file execution.
Keep Python runtime process execution here.
"""

import os
import subprocess
import sys
import time

MAX_ARTIFACT_BYTES = 256 * 1024
ENTRY_FILE_WAIT_MS = 5000
ENTRY_FILE_WAIT_INTERVAL_MS = 100

def handler(event, context):
    started_at = time.time()
    try:
        if event.get("runtime") != "python":
            raise ValueError("sandbox-python only supports python requests")

        entry_path = event.get("entryPath", "")
        if not entry_path.endswith(".py"):
            raise ValueError("python sandbox only executes .py files")

        workspace_root = resolve_workspace_root(event.get("workspaceRoot"), event.get("namespace", ""))
        file_path = resolve_workspace_path(workspace_root, entry_path)
        await_file(file_path)
        if not os.path.isfile(file_path):
            raise FileNotFoundError(file_path)
        before = snapshot_workspace(workspace_root)
        result = run_python_file(
            file_path,
            workspace_root,
            event.get("args") or [],
            int(event.get("timeoutSeconds", 30)),
            int(event.get("outputLimitBytes", 64 * 1024)),
        )
        result.update({
            "runtime": "python",
            "artifacts": collect_changed_artifacts(workspace_root, before),
            "durationMs": int((time.time() - started_at) * 1000),
        })
        return result
    except Exception as err:
        return {
            "ok": False,
            "runtime": "python",
            "exitCode": None,
            "stdout": "",
            "stderr": str(err),
            "durationMs": int((time.time() - started_at) * 1000),
        }

def await_file(file_path: str) -> None:
    deadline = time.time() * 1000 + ENTRY_FILE_WAIT_MS
    last_error = None

    while time.time() * 1000 < deadline:
        try:
            if os.path.isfile(file_path):
                return
        except Exception as err:
            last_error = err
            time.sleep(ENTRY_FILE_WAIT_INTERVAL_MS / 1000)
            continue

    if last_error:
        raise last_error
    if not os.path.isfile(file_path):
        raise FileNotFoundError(file_path)


def run_python_file(file_path, cwd, args, timeout_seconds, output_limit_bytes):
    try:
        completed = subprocess.run(
            [sys.executable, "-I", "-S", file_path, *args],
            cwd=cwd,
            env={
                "PATH": "/usr/local/bin:/usr/bin:/bin",
                "HOME": "/tmp",
                "TMPDIR": "/tmp",
                "PYTHONPATH": "",
            },
            capture_output=True,
            timeout=timeout_seconds,
            check=False,
        )
        stdout, stdout_truncated = truncate_output(completed.stdout, output_limit_bytes)
        stderr, stderr_truncated = truncate_output(completed.stderr, output_limit_bytes)
        return {
            "ok": completed.returncode == 0,
            "exitCode": completed.returncode,
            "stdout": stdout,
            "stderr": stderr,
            "truncated": stdout_truncated or stderr_truncated,
        }
    except subprocess.TimeoutExpired as err:
        stdout, stdout_truncated = truncate_output(err.stdout or b"", output_limit_bytes)
        stderr, stderr_truncated = truncate_output(err.stderr or b"", output_limit_bytes)
        return {
            "ok": False,
            "exitCode": None,
            "stdout": stdout,
            "stderr": append_timeout(stderr, timeout_seconds),
            "timedOut": True,
            "truncated": stdout_truncated or stderr_truncated,
        }


def resolve_workspace_root(root, namespace):
    assert_safe_namespace(namespace)
    return os.path.abspath(os.path.join(root or "/mnt/workspaces", namespace))


def resolve_workspace_path(workspace_root, entry_path):
    normalized_entry = entry_path[1:] if entry_path.startswith("/") else entry_path
    resolved = os.path.abspath(os.path.join(workspace_root, normalized_entry))
    common = os.path.commonpath([workspace_root, resolved])
    if common != workspace_root or resolved == workspace_root:
        raise ValueError("Invalid entry path: resolved outside workspace root")
    return resolved


def snapshot_workspace(root):
    files = {}
    for path in iter_workspace_files(root):
        stats = os.stat(path)
        files[os.path.relpath(path, root)] = {
            "mtime_ns": stats.st_mtime_ns,
            "size": stats.st_size,
        }
    return files


def collect_changed_artifacts(root, before):
    import base64

    artifacts = []
    for path in iter_workspace_files(root):
        relative_path = os.path.relpath(path, root)
        stats = os.stat(path)
        previous = before.get(relative_path)
        if previous and previous["mtime_ns"] == stats.st_mtime_ns and previous["size"] == stats.st_size:
            continue
        if stats.st_size > MAX_ARTIFACT_BYTES:
            continue

        with open(path, "rb") as file:
            content = file.read()

        artifacts.append({
            "kind": "file",
            "path": f"/{relative_path}",
            "mediaType": "application/octet-stream",
            "dataBase64": base64.b64encode(content).decode("ascii"),
            "metadata": {
                "size": stats.st_size,
            },
        })
    return artifacts


def iter_workspace_files(root):
    for current_root, dirs, files in os.walk(root):
        dirs[:] = [name for name in dirs if name not in (".", "..")]
        for file_name in files:
            path = os.path.join(current_root, file_name)
            if os.path.isfile(path):
                yield path


def assert_safe_namespace(namespace):
    import re

    if not re.fullmatch(r"fs-[a-f0-9]{40}", namespace):
        raise ValueError("Invalid workspace namespace")


def truncate_output(value, limit):
    if len(value) <= limit:
        return value.decode("utf-8", errors="replace"), False

    return value[:limit].decode("utf-8", errors="replace") + "\n[output truncated]", True


def append_timeout(stderr, timeout_seconds):
    timeout_text = f"Timed out after {timeout_seconds}s"
    return "\n".join(part for part in [stderr, timeout_text] if part)
